src/persist/src/indexed/encoding.rs

```diff
@@ -269,7 +271,83 @@ impl Default for BlobMeta {
     }
 }
 
+struct ExtendWriteAdapter<'e, E>(&'e mut E);
+
+impl<'e, E: for<'a> Extend<&'a u8>> io::Write for ExtendWriteAdapter<'e, E> {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, io::Error> {
+        self.0.extend(buf);
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> Result<(), io::Error> {
+        Ok(())
+    }
+}
+
+// The encoding of BlobMeta is:
+// ```
+// buf[0]          = <BlobMeta::CURRENT_VERSION>
+// buf[1..3]       = <BlobMeta::MAGIC>
+// buf[3..3+N]     = <the N-byte Abomonation serialization of BlobMeta>
+// buf[3+N..3+N+2] = <BlobMeta::MAGIC>
+// ```
+//
+// This will allow us to gracefully detect changes at startup and react to them.
+// MZ is not (yet) a system of record, so in the short-term (persisting
+// mz_metrics) we are free to simply delete the data. In the medium-term (fast
+// restarts for kafka sources), we'll make an effort to decode old data. In the
+// long-term (1.0), we'll have backward-compatibility guarantees.
+impl Codec for BlobMeta {
+    fn codec_name() -> &'static str {
+        "BlobMeta"
+    }
+
+    fn size_hint(&self) -> usize {
+        0
+    }
+
+    fn encode<E: for<'a> Extend<&'a u8>>(&self, buf: &mut E) {
+        buf.extend(&[BlobMeta::CURRENT_VERSION]);
+        buf.extend(BlobMeta::MAGIC);
+        unsafe { abomonation::encode(self, &mut ExtendWriteAdapter(buf)) }
+            .expect("write to ExtendWriteAdapter is infallible");
+        buf.extend(BlobMeta::MAGIC);
+    }
+
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
+        let version_pos = 0..1;
+        let begin_magic_pos = version_pos.end..version_pos.end + BlobMeta::MAGIC.len();
+        // Avoid panicking when buf is short.
+        let end_magic_start = if begin_magic_pos.end + BlobMeta::MAGIC.len() > buf.len() {
+            begin_magic_pos.end
+        } else {
+            buf.len() - BlobMeta::MAGIC.len()
+        };
+        let end_magic_pos = end_magic_start..buf.len();
+        match buf.get(version_pos.start) {
+            Some(&BlobMeta::CURRENT_VERSION) => {}
+            Some(x) => return Err(format!("unsupported version: {}", x)),
+            None => return Err("unknown version".into()),
+        }
+        match buf.get(begin_magic_pos.clone()) {
+            Some(BlobMeta::MAGIC) => {}
+            _ => return Err("bad magic".into()),
+        }
+        match buf.get(end_magic_pos.clone()) {
+            Some(BlobMeta::MAGIC) => {}
+            _ => return Err("bad magic".into()),
+        }
+        let buf = &buf[begin_magic_pos.end..end_magic_pos.start];
+        let meta: Abomonated<BlobMeta, Vec<u8>> =
+            unsafe { Abomonated::new(buf.to_owned()) }.ok_or_else(|| "invalid meta")?;
```

Comment:
nit: this does one more allocation, compared to the previous code, I think. But I'm guessing it's alright because this is not performance sensitive.

Reply:
Yeah I was sad about this too :(. I suppose I could have only put the version and magic at the end, which would have let me pass in a vec and truncate it. But as you mention, this isn't performance sensitive (we only decode META one time ever at startup). Plus, we're going to replace abom at some point with a protocol that has a backward/forward compatibility story (protobuf/capnp/json/etc), and all of those let you decode non-owned bytes.